File: Solutions_Easy/0125_ValidPalindrome.py
import re


class Solution:
    def isPalindrome(self, s: str) -> bool:
        # Uppercase -> Lowercase
        s = s.lower()
        # Removes all non-alphanumeric chars
        s = re.sub(r"[^a-z0-9]", "", s)

        for s1, s2 in zip(s[: len(s) // 2], s[::-1]):
            if s1 != s2:
                return False
        return True

File: Solutions_Easy/0125_ValidPalindrome.cpp
class Solution {
public:
    bool isPalindrome(std::string s) {
        std::string cleaned;
        for (char c : s) {
            if (isalnum(c)) {          // Filter non-alphanumeric chars
                cleaned += tolower(c); // Palindrome doesn't care about upper/lowercase difference
            }
        }

        int n = cleaned.size();
        for (int i = 0; i < n / 2; ++i) {
            if (cleaned[i] != cleaned[n - 1 - i]) {
                return false;
            }
        }

        return true;
    }
};